### PR TITLE
Feature project

### PR DESCRIPTION
**Pull request by sjcco**
PROJECT: ASSOCIATIONS

- Setup and Sign In
- Basic Events
- Event Attendance

### CODE REVIEW

**Review by mke2111 — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coder, 👋🏽 
**This is a great project you have going on, well-done lad**. 

### Well-done 💯 
    1. The app runs just fine
    2. the linter checks pass

**However, there are a few sections to improve before your project is approved.** 🔴 
<hr>

### REQUIRED

- [x] In the readme file, add the following commands before starting the server.
`bundle install` - this command obtains all the dependencies required by your app
`rails db:migrate` - this creates the tables within the dB.

- [x] Remove the `bin` from your seed and server commands. They need to run from the root directory.
- [x] Add `rails webpacker:install` to install the configured webpack file that helps run your app. Currently, the app won't run successfully because this file is missing
- [x] Currently, a user is able to subscribe/attend a previous event. This shouldn't be the expected behaviour. You may choose to hide the attend event button if the event is in the past.
- [x] models have unit tests for validations and associations (Rspec) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/unit_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [x] authentication have has integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [x] event managment features have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))

- [x] Check in-line comments for more sections to improve



#

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

<hr>

Mention @mke2111 in this thread and I'll be glad to help. 😄 
### Best regards
Slack 🆘  Roy Mukuye
Twitter 🐦  [Roy Mukuye](https://twitter.com/Roymkenya)
LinkedIn 🔗  [Roy Mukuye](https://www.linkedin.com/in/roy-mukuye/)

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
+class EventsController < ApplicationController
+  before_action :require_login, only: [:show]
+
+  def index
+    @events = Event.all
+  end
```
Comment (mke2111) [inline app/controllers/events_controller.rb:8]: - [ ] The index action lists past events, i.e. @past_events = Event.previous_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin)

- [ ] The index action lists ucomping events, i.e. @upcoming_events = Event.upcoming_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from Odin)

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    @user = User.new(user_params)
+    @user.save
+    session[:user_id] = @user.id
+    flash.notice = "User #{@user.name} has succesfully signed in"
+    redirect_to root_path
+  end
+
+  def show
+    @user = User.find(params[:id])
+  end
```
Comment (mke2111) [inline app/controllers/users_controller.rb:23]: - [ ] The show action lists events created by user, i.e. @created_events = current_user.events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from Odin)
- [ ] The show action lists events previoulsy attendented by user using event model scope, i.e. @past_events = current_user.attended_events.previous ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from Odin)
- [ ] The show action lists upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from Odin)

File: app/views/events/show.html.erb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    </div>
+    <div class="info">
+      <p><i class="fas fa-user-astronaut"></i> <%= @event.creator.name %></p>
+      <p><i class="far fa-calendar-alt"></i> <%= @event.starting_date_time.strftime("%B %-d, %Y at %l:%M %p") %></p>
+      <p><i class="far fa-calendar-alt"></i> <%= @event.ending_date_time.strftime("%B %-d, %Y at %l:%M %p") %></p>
+      <p><i class="fas fa-map-marked-alt"></i> <%= @event.location %></p>
+    </div>
+
+  </div>
+  <% if @event.attendees.include?(current_user) %>
```
Comment (mke2111) [inline app/views/events/show.html.erb:16]: OPTIONAL

- [ ] It is a good practice to have zero logic in the views directory. Create methods in the helpers directory and call the methods in the views directory instead.
Read this [article](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) for a head-start. Another great piece to read is this [article](https://brandonhilkert.com/blog/refactoring-logic-from-a-rails-view/)

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,16 +1,67 @@
...
+</head>
+
+<body>
+  <div class="container">
+    <nav>
+      <%= link_to root_path do %> <div>
+        <%= image_tag "logo.png" %> <p>Bati-Events</p>
+      </div> <% end %>
+      <ul>
+        <% if user_signed_in? %>
```
Comment (mke2111) [inline app/views/layouts/application.html.erb:23]: Another piece of logic statement

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -1,16 +1,67 @@
...
+        <li>
+          <%= link_to 'Sign up', signup_path %>
+        </li>
+
+        <% end %>
+
+      </ul>
+    </nav>
+  </div>
+  <% if flash[:notice] %>
```
Comment (mke2111) [inline app/views/layouts/application.html.erb:50]: Another logic statement

File: app/controllers/events_controller.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
+class EventsController < ApplicationController
+  before_action :require_login, only: [:show]
+
+  def index
+    @events = Event.all
```
Comment (afanimartin) [inline app/controllers/events_controller.rb:5]: - Add @past_events = Event.previous_events and @upcoming_events = Event.upcoming_events

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+  def create
+    @user = User.new(user_params)
+    @user.save
+    session[:user_id] = @user.id
+    flash.notice = "User #{@user.name} has succesfully signed in"
+    redirect_to root_path
+  end
+
+  def show
+    @user = User.find(params[:id])
```
Comment (afanimartin) [inline app/controllers/users_controller.rb:19]: - The show action lists events created by user, i.e. @created_events = current_user.events (screenshot from Odin)
- The show action lists events previoulsy attendented by user using event model scope, i.e. @past_events = current_user.attended_events.previous (screenshot from Odin)
- The show action lists upcoming events previously attended by user using event model scope, i.e. @upcoming_events = current_user.attended_events.upcoming (screenshot from Odin)

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 🙏🏿 🙏🏿 
Nice work making some the requested changes.
- A user is now unable to attend past events
- Integration tests have been added

Let's make some of the suggested changes.
- Check my comments for suggestions


### Martin Afani
#### Slack: whiz25

**Review by voscarmv — APPROVED** [review]
# Project status: Approved 💯 ✨ 😎 👍 
- Good work getting the project in full working order.
### You may merge this branch and move forward ⛙🛣️ 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻
# Happy Coding! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**